File: Top_100_Liked_Questions/300._Longest_Increasing_Subsequence/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    // O(n) space
    if (nums.length <= 1) {
        return nums.length;
    }
    
    let mem = new Array(nums.length).fill(0);
    mem[0] = 1;
    let maxAns = 1;
    
    for (let i=0; i<mem.length; i++) {
        let maxVal = 0;
        for (let j=0; j<i; j++) {
            if (nums[i] > nums[j]) {
                maxVal = Math.max(maxVal, mem[j]);
            }
        }
        mem[i] = maxVal + 1;
        maxAns = Math.max(maxAns, mem[i]);
    }
    
    return maxAns;
}


var lengthOfLIS3 = function(nums) {
    // dp with memoization - O(n^2) space
    if (nums.length <= 1) {
        return nums.length;
    }
    
    nums.unshift(-Infinity);
    
    let mem = new Array(nums.length+1);
    for (let i=0; i<mem.length; i++) {
        mem[i] = new Array(nums.length).fill(0);
    }
    
    for (let i=nums.length-1; i>0; i--) {
        for (let j=0; j<i; j++) {
            if (nums[i] > nums[j]) {
                mem[i][j] = Math.max(1 + mem[i+1][i], mem[i+1][j]);
            } else {
                mem[i][j] = mem[i+1][j];
            }
        }
    }

    return mem[1][0];
}


var lengthOfLIS2 = function(nums) {
    // recursion
    nums = [-Infinity, ...nums];
    let LTS = function(i, j) {
        if (i >= nums.length) {
            return 0;
        }
        
        if (nums[i] > nums[j]) {
            return Math.max(1 + LTS(i+1, i), LTS(i+1, j));
        } else {
            return LTS(i+1, j);
        }
    }
    
    return LTS(1, 0);
};

File: Top_100_Liked_Questions/300._Longest_Increasing_Subsequence/solution.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return len(nums)
        
        mem = [0] * (len(nums))
        mem[0] = 1
        maxAns = 1
        for i in range(1, len(nums)):
            maxVal = 0
            for j in range(i):
                if nums[i] > nums[j]:
                    maxVal = max(maxVal, mem[j])    
            
            mem[i] = maxVal + 1
            maxAns = max(maxAns, mem[i])
        
        return maxAns
    
    
    def lengthOfLIS3(self, nums: List[int]) -> int:
        if len(nums) <= 1:
            return len(nums)
        
        nums.insert(0, float('-inf'))
        mem = [[0] * len(nums) for _ in range(len(nums)+1)]
        
        for i in range(len(nums)-1, 0, -1):
            for j in range(i):
                if nums[i] > nums[j]:
                    mem[i][j] = max(mem[i+1][j], 1 + mem[i+1][i])
                else:
                    mem[i][j] = mem[i+1][j]
                    
        return mem[1][0]
    
    
    def lengthOfLIS2(self, nums: List[int]) -> int:
        # recursive dp
        nums.insert(0, float('-inf'))
        
        def LIS(i, j):
            if i >= len(nums):
                return 0
            
            if nums[i] > nums[j]:
                return max(LIS(i+1, j), 1 + LIS(i+1, i))
            else:
                return LIS(i+1, j)
        
        return LIS(1,0)